src/mscorlib/shared/System/IO/PathHelper.Windows.cs

```diff
@@ -42,14 +42,31 @@ internal static string Normalize(string path)
             return result;
         }
 
-        private static void GetFullPathName(string path, ref ValueStringBuilder builder)
+        internal static string Normalize(ReadOnlySpan<char> path)
+        {
+            Span<char> initialBuffer = stackalloc char[PathInternal.MaxShortPath];
+            ValueStringBuilder builder = new ValueStringBuilder(initialBuffer);
+
+            // Get the full path
+            GetFullPathName(path, ref builder);
+
+            string result = builder.AsSpan().Contains('~')
+                ? TryExpandShortFileName(ref builder, originalPath: null)
+                : builder.ToString();
+
+            // Clear the buffer
+            builder.Dispose();
+            return result;
+        }
+
+        internal static void GetFullPathName(ReadOnlySpan<char> path, ref ValueStringBuilder builder)
         {
             // If the string starts with an extended prefix we would need to remove it from the path before we call GetFullPathName as
             // it doesn't root extended paths correctly. We don't currently resolve extended paths, so we'll just assert here.
             Debug.Assert(PathInternal.IsPartiallyQualified(path) || !PathInternal.IsExtended(path));
 
             uint result = 0;
-            while ((result = Interop.Kernel32.GetFullPathNameW(path, (uint)builder.Capacity, ref builder.GetPinnableReference(), IntPtr.Zero)) > builder.Capacity)
+            while ((result = Interop.Kernel32.GetFullPathNameW(ref MemoryMarshal.GetReference(path), (uint)builder.Capacity, ref builder.GetPinnableReference(), IntPtr.Zero)) > builder.Capacity)
```

Comment:
What guarantees that the path is zero terminated?

Reply:
I think this may be going overboard with trying to avoid allocations. Bad tradeoff between simplicity/maintainability and performance.

Reply:
The APIs we call null terminate. I can add an unsafe assert here.

Reply:
You are right that it happens to work today because of there is zero termination done 3 hops away, but it very hard to follow.

The regular contract for Span<char> is to not be zero terminated. There should be comment next to every Span local and argument that is assumed to be zero terminated.

Reply:
I'll make sure the comment is in there. It raises an interesting question about interop and ValueStringBuilder/string/Span. We don't have a great way to represent a null terminated Span. Should we have ToSpanWithNull()?

Reply:
ToSpanWithNull would be a way to get zero a null terminated span, but it is not a way to represent a zero terminated span.

I agree that it would be nice to have a set of conventions on how to represent zero terminated spans.

We are doing `Append('\0')` and `GetPinnableReference` today that I do not think is too bad to get a zero terminated string. Not sure whether `ToSpanWithNull` method would be an improvement.

Reply:
Discussed this some with @KrzysztofCwalina and opened https://github.com/dotnet/corefx/issues/28040 for getting the terminator from the string. I'm updating the pull here to add the assert in such a manner.